### PR TITLE
Change fastcdc to a better and simpler algorithm.

### PR DESCRIPTION
**Pull request by dbaarda**
This changes the chunking algorithm from "normalized chunking" to simple "exponential chunking", changes the hash criteria from `hash&mask` to `hash<=threshold`, and changes the gear rollsum to use a multiply instead of a
table lookup.

These are all ideas taken from testing and analysis done at;

* https://github.com/dbaarda/rollsum-chunking/blob/master/RESULTS.rst
* https://github.com/dbaarda/rollsum-tests/blob/master/RESULTS.rst

### CODE REVIEW

**Comment by dbaarda** [comment]
Note this is to show the concept and for testing. I expect this will need some tweaking if having a normal distribution of chunk sizes is genuinely a requirement, even if it does hurt deduplication.

File: fastcdc/fastcdc.h [hunk 1]
```diff
@@ -29,8 +29,11 @@
 namespace cdc_ft {
 namespace fastcdc {
 
-static constexpr uint32_t default_mask_stages = 7;
-static constexpr uint32_t default_mask_lshift = 3;
+// These are recommended and widely used multipliers for LCGs from
+// http://citeseer.ist.psu.edu/132363.html. They replace the gear lookup
```
Comment (ljusten) [inline fastcdc/fastcdc.h:33]: This site doesn't seem to exist anymore.

File: fastcdc/fastcdc.h [hunk 2]
```diff
@@ -57,53 +60,53 @@ using ChunkFoundHandler = std::function<void(const uint8_t* data, size_t len)>;
 // Implements a very fast content-defined chunking algorithm.
 //
 // FastCDC [1] identifies chunk boundaries based on a simple yet efficient
```
Comment (ljusten) [inline fastcdc/fastcdc.h:62]: If we go with this implementation, it doesn't make much sense anymore to call it fastcdc, I guess.

File: fastcdc/fastcdc.h [hunk 3]
```diff
@@ -240,180 +162,38 @@ class ChunkerTmpl {
...
+    for (/*empty*/; i < len; ++i) {
+      if (hash <= kthreshold_) {
+        return i;
       }
+      hash = ((hash << 1) + data[i]) * kmult;
     }
     return i;
   }
 
+  static constexpr int khashbits_ = sizeof(T) * 8;
```
Comment (ljusten) [inline fastcdc/fastcdc.h:181]: Nit: Maybe use size_t instead of int as it's compared to a size_t above.

**Review by ljusten — CHANGES_REQUESTED** [review]

File: fastcdc/fastcdc.h (hunk 2 above)
Comment (dbaarda) [reply fastcdc/fastcdc.h:62]: Yeah, probably correct. I wanted to minimize the changes for now. I think git can also get a bit confused merging things after directory renames (or is that old news?).

File: fastcdc/fastcdc.h (hunk 1 above)
Comment (dbaarda) [reply fastcdc/fastcdc.h:33]: Sigh! It seems pretty hard to find stable links to academic papers that aren't locked behind paywalls. Perhaps This one is better?

https://www.ams.org/journals/mcom/1999-68-225/S0025-5718-99-00996-5/S0025-5718-99-00996-5.pdf

File: fastcdc/fastcdc.h (hunk 3 above)
Comment (dbaarda) [reply fastcdc/fastcdc.h:181]: Done.

**Comment by dbaarda** [comment]
So there are a bunch of things in this that I'd be interested in seeing tested against whatever test-data you have. I don't really have anything setup to test this, I have no idea what kind of data it's normally used with, and don't have a windows box to test with.

My testing of the algorithm suggests it should give noticeably better deduplication, and should be faster. Some questions are;

1. How important is the normal distribution of chunk sizes? My testing showed that normalizing the chunk sizes significantly hurts deduplication, so is it worth it? More on alternatives below...
2. Is using a mult instead of a table lookup for the gear hash faster or slower? A table lookup is normally a little faster but eats 2K of L1 cache, so it's often slower in the big-picture of a whole application that also needs L1 cache.
3. Will using a 32bit Gear hash be just as good at deduplication but faster? Most other implementations just use a 32bit hash and seem fine with it. Note the testing using a 32bit lookup table was using it with a 64bit hash, which wouldn't have given good distribution of entropy over the hash, and bitmask hash criteria would have lost even more. If you are going to use a 32bit lookup table, you should use a 32bit hash and use `hash<=threshold" to get the most entropy out of it.

If there is a need for tightly constrained chunk sizes, my testing showed that using a simple exponential chunker with `min_size=avg_size/2,max_size=avg_size*2` still gives better deduplication than a normalizing chunker with a larger `max_size=avg_size*4`. For even better deduplication or an even tighter size range with `max_size=1.5*avg_size`, use "Regression Chunking", which would be easy to add, and gives an almost flat chunk size distribution.

If you have any ideas of suggestions I'm willing to play with it.

**Comment by ljusten** [comment]
1. Since the tools were intended for Stadia, we always tested against games. For cdc_rsync we had a bunch of dev builds of some game over the course of 2 months or so, which we uploaded consecutively and measured timing. This is a good real world use case and measures deduplication, compression bandwidth, IO performance etc. For cdc_stream we started a bunch of games on Stadia and measured the time it took to reach the menu. This only really tests file IO time and streaming bandwidth, not deduplication. The ability to repeat the tests are crumbling as Stadia is being dismantled. We still have the data for the cdc_rsync test, but right now we can't compile for the cloud instances anymore (would be a bit of work to port the Bazel toolchain back), so we'd have test using different hardware. Still, comparing the old to the new algo would be possible. The cdc_stream test is a bit harder to repro. I'm not sure whether the data is still available. I also didn't create the tests myself and Christian is no longer actively working on this project. We can test streaming stuff like InfiltratorDemo, though, and just measure time with something like `cat somefile > /dev/null`. This should give a good impression how the chunk size distribution impacts streaming performance.
2. Good question. On modern CPUs I wouldn't be surprised if it was faster as it's more important to mitigate memory access nowadays. I can run some benchmarks if I find the time.
3. We measured marginally better dedupe rates. CPUs should handle 64 ops pretty much as fast as 32 bit ops, so as long as there isn't much more memory overhead, the performance difference should be negligible. We haven't measured that, though.
If you want to play with it, you could compare the performance of the chunker. You could probably just download some random game or take a larger Windows file, but I don't think the file type matters much. Without a Windows box it's hard to test the actual tools, though you could get a gWindows cloudtop and apply for a Bit9 exception, so you can run self-compiled binaries. That takes about a day.

**Comment by dbaarda** [comment]
I've fixed up the indexer for the fastcdc changes so those tests now pass, but the windows manifest_updater_test is failing. It looks to me like it has hard-coded expectations about the chunking in some tests.

File: fastcdc/fastcdc.h (hunk 1 above)
Comment (dbaarda) [reply fastcdc/fastcdc.h]: I've updated this to the link that seems to work.

**Comment by dbaarda** [comment]
I now have 4 different versions of this in different branches;

* faster_cdc : this #79 PR using a modified gear with a multiplier.
* faster_cdc_rcn: #84 PR the same as faster_cdc with regression chunking added.
* faster_cdc_table : the #79 PR with gear switched back to using a lookup table.
* faster_cdc_rcn_table: The #84 PR with gear switched back to using a lookup table.

I haven't made PR's for the last 2 as I didn't want to flood the PR system with candidates that will get abandoned. If anyone else really wants PRs for these let me know.

So far in testing it seems the lookup table versions are much faster with maybe a tiny hit in deduplication. There is a possibility that this speed benefit goes away in the wider context of an application actually doing more than just chunking and needing the L1 cache, but maybe not. The addition of rcn doesn't seem to impact speed much at all for typical min=avg/2, max=avg*4 sizes you would use with a normal exponential chunker, but you can also set max=avg*2 or even max=avg*1.5 for a tightly constrained chunk size distribution and still get very good deduplication  for a small speed hit.

Compared to the current main branch the faster_cdc_rcn_table version is much simpler, about the same speed, and gives much better deduplication.

**Comment by dbaarda** [comment]
Since testing has narrowed down on the best combination of gear with lookup table and regression chunking, I'm updating this branch with a clean rebased commit history with the best current candidate.

We still need to fix the failing windows test though.

**Comment by dbaarda** [comment]
This now passes all tests and chunker settings for cdc_stream have been updated to take advantage of the tighter max limit it can handle. It gives better deduplication at about the same chunking speed, which translates to noticably faster transfers/streaming. The code is also much simpler.

I think this is ready to merge.

File: fastcdc/fastcdc.h (hunk 2 above)
Comment (dbaarda) [reply fastcdc/fastcdc.h:59]: It's a pretty disruptive change to rename it... it's all over the codebase.

Personally I'm OK with leaving it as it is. Although it's not the same as the original fastcdc, the research and changes I did started with investigating it, and the changes are well documented in comments.

File: fastcdc/fastcdc.h [hunk 4]
```diff
@@ -239,30 +160,42 @@ class ChunkerTmpl {
...
+    // mask to an empty bitmask (match any hash).
+    size_t rc_len = len;
+    T rc_mask = 0;
+
     // Init hash to all 1's to avoid zero-length chunks with min_size=0.
-    uint64_t hash = UINT64_MAX;
-    // Skip the first min_size bytes, but "warm up" the rolling hash for 64
-    // rounds to make sure the 64-bit hash has gathered full "content history".
-    size_t i = cfg_.min_size > 64 ? cfg_.min_size - 64 : 0;
+    T hash = (T)-1;
```
Comment (ljusten) [inline fastcdc/fastcdc.h:169]: Nit: Same here

File: fastcdc/fastcdc.h [hunk 5]
```diff
@@ -239,30 +160,42 @@ class ChunkerTmpl {
...
-        hash = (hash << 1) + gear[data[i]];
       }
+      hash = (hash << 1) + gear[data[i]];
     }
-    return i;
+    // Return best regression point we found or the end if it's better.
+    return (hash & rc_mask) ? rc_len : i;
   }
 
+  static constexpr size_t khashbits_ = sizeof(T) * 8;
```
Comment (ljusten) [inline fastcdc/fastcdc.h:195]: Nit: kHashBits as per style guide

File: fastcdc/fastcdc.h [hunk 6]
```diff
@@ -239,30 +160,42 @@ class ChunkerTmpl {
...
-      uint64_t mask = stages_[stg].mask;
-      size_t barrier = std::min(len, stages_[stg].barrier);
-      for (/*empty*/; i < barrier; ++i) {
-        if (!(hash & mask)) {
+    for (/*empty*/; i < len; ++i) {
+      if (!(hash & rc_mask)) {
+        if (hash <= kthreshold_) {
+          // This hash matches the target length hash criteria, return it.
           return i;
+        } else {
```
Comment (ljusten) [inline fastcdc/fastcdc.h:181]: Nit: Remove the 'else', it's not needed since the code above returns

File: fastcdc/fastcdc.h [hunk 7]
```diff
@@ -239,30 +160,42 @@ class ChunkerTmpl {
...
+          // This is a better regression point. Set it as the new rc_len and
+          // update rc_mask to check as many MSBits as this hash would pass.
+          rc_len = i;
+          for (rc_mask = (T)-1; hash & rc_mask; rc_mask <<= 1)
+            ;
```
Comment (ljusten) [inline fastcdc/fastcdc.h:186]: Nit: Maybe { /* empty */ } for better readability. Or
rc_mask = std::numeric_limits<T>::max();
while (hash & rc_mask) rc_mask <<= 1;

File: fastcdc/fastcdc.h [hunk 8]
```diff
@@ -239,30 +160,42 @@ class ChunkerTmpl {
...
-    return i;
+    // Return best regression point we found or the end if it's better.
+    return (hash & rc_mask) ? rc_len : i;
   }
 
+  static constexpr size_t khashbits_ = sizeof(T) * 8;
   const Config cfg_;
   const ChunkFoundHandler handler_;
-  MaskStage stages_[mask_stages];
+  T kthreshold_;
```
Comment (ljusten) [inline fastcdc/fastcdc.h:198]: Nit: We generally use the k notation only if the variable is actually a const. Maybe just threshold_ or make it const T (which you can do if you compute it after the : in the constructor, though that's a bit ugly).

**Review by ljusten — APPROVED** [review]
Looks good, thanks! Just some style and other minor nits.

File: fastcdc/fastcdc.h [hunk 9]
```diff
@@ -57,53 +57,56 @@ using ChunkFoundHandler = std::function<void(const uint8_t* data, size_t len)>;
...
-                  "mask_stages must be between 1 and 64");
-    static_assert(mask_lshift > 0 && mask_lshift <= 31,
-                  "mask_lshift must be between 1 and 31");
+    assert(cfg_.avg_size >= 1);
     assert(cfg_.min_size <= cfg_.avg_size);
     assert(cfg_.avg_size <= cfg_.max_size);
 
+    // Calculate the threshold the hash must be <= to for a 1/(avg-min+1)
+    // chance of a chunk boundary.
+    kthreshold_ = (T)(-1) / (cfg_.avg_size - cfg_.min_size + 1);
```
[the remark this answers is not in the document]
Comment (dbaarda) [reply fastcdc/fastcdc.h]: Done.

File: fastcdc/fastcdc.h (hunk 4 above)
Comment (dbaarda) [reply fastcdc/fastcdc.h]: Done.

File: fastcdc/fastcdc.h (hunk 5 above)
Comment (dbaarda) [reply fastcdc/fastcdc.h]: Done.

File: fastcdc/fastcdc.h (hunk 6 above)
Comment (dbaarda) [reply fastcdc/fastcdc.h]: Done.

File: fastcdc/fastcdc.h (hunk 7 above)
Comment (dbaarda) [reply fastcdc/fastcdc.h]: Done.

**Pull request merged**